src/LineBlockTracer.cpp

```diff
@@ -49,6 +50,160 @@ bool cLineBlockTracer::Trace(cWorld & a_World, cBlockTracer::cCallbacks & a_Call
 
 
 
+bool cLineBlockTracer::LineOfSightTrace(cWorld & a_World, const Vector3d & a_Start, const Vector3d & a_End, eLineOfSight a_Sight)
+{
+	switch (a_Sight)
+	{
+		case losAir:
+		{
+			static class LineOfSightCallbacksAir:
+				public cLineBlockTracer::cCallbacks
+			{
+			public:
+				LineOfSightCallbacksAir() {}
+
+				virtual bool OnNextBlock(int a_BlockX, int a_BlockY, int a_BlockZ, BLOCKTYPE a_BlockType, NIBBLETYPE a_BlockMeta, eBlockFace a_EntryFace) override
+				{
+					if (a_BlockType != E_BLOCK_AIR)
+					{
+						return true;  // Abort tracing
+					}
+					return false;
+				}
+			} callbacksAir;
+			return Trace(a_World, callbacksAir, a_Start, a_End);
+		}
+
+		case losAirWater:
+		{
+			static class LineOfSightCallbacksAirWater:
+				public cLineBlockTracer::cCallbacks
+			{
+			public:
+				LineOfSightCallbacksAirWater() {}
+
+				virtual bool OnNextBlock(int a_BlockX, int a_BlockY, int a_BlockZ, BLOCKTYPE a_BlockType, NIBBLETYPE a_BlockMeta, eBlockFace a_EntryFace) override
```

Comment:
Templating might be the next step, if we find out this piece of code needs optimizing.

The callbacks are implemented as a class because they can have multiple entrypoints (`OnNextBlock`, `OnOutOfWorld`, `OnNoMoreHits` etc.), I don't think that can be reasonably achieved with closures.